TW edits: Workers tutorial (#13199)
PrefectHQ/prefect: docs/tutorial/workers.md

# Workers

This part of the tutorial creates a Docker work pool with workers that can handle more complex infrastructure management and monitoring.

## Prerequisites

> * Install [Docker](https://docs.docker.com/engine/install/).

Reviewer: ```suggestion
* Install [Docker](https://docs.docker.com/engine/install/) and ensure it is running.
```

    worker --> |creates| flow_run_infra_2
```

## Create a work pool

1. Create a Docker work pool named `my-docker-pool`:

    <div class="terminal">

    ```bash
    prefect work-pool create --type docker my-docker-pool
    ```

    </div>

    Using the **Docker** work pool type means that all work sent to this work pool runs within a dedicated Docker container using a Docker client available to the worker.

1. Verify its creation:

    <div class="terminal">

    ```bash
    prefect work-pool ls
    ```

    </div>

1. Confirm its visibility on the Prefect UI's **Work Pools** tab (red status indicates it's not ready yet).

## Start a worker

Workers monitor the Prefect API for scheduled flow runs and initiate execution on your chosen infrastructure.

> Start a local worker for `my-docker-pool`:

Reviewer: ```suggestion
Start a worker for `my-docker-pool`:
```

</div>

> The worker appears in the UI's **Workers** tab (Work Pools page), with a recent last polled date to check for any scheduled flow runs it should pick up and then submit for execution. Your work pool's status should now be `Ready`.

Reviewer: ```suggestion
The worker's status appears in the UI's Work Pools page, on the **Workers** tab. Your work pool's status should now be `Ready`.
```

> This local worker terminates if you close the terminal session. In production, workers should run as a [daemonized or managed process](/guides/deployment/daemonize/).

Reviewer: ```suggestion
The worker terminates if you close the terminal session. In production, a worker should generally run as a [daemonized process](/guides/deployment/daemonize/).
```

## Create the deployment

Deploy your flow to `my-docker-pool`:

1. Make the following changes to your `repo_info.py` file:

    * Change `flow.serve` to `flow.deploy`.
    * Specify your target work pool using `flow.deploy`.

>     * Set the name for the Docker image

Reviewer 2: ```suggestion
    * Set the name for the Docker image.
```

        get_repo_info.deploy(
            name="my-first-deployment", 
            work_pool_name="my-docker-pool", 
            image="my-first-deployment-image:tutorial",
            push=False
        )
    ```

    For this tutorial, your Docker worker is running on your machine, so you don't need to push the image built by `flow.deploy` to a registry. When your worker is running on a remote machine, you need to [push the image to a registry that the worker can access](/guides/prefect-deploy/#creating-work-pool-based-deployments-with-deploy).

1. Deploy your flow to the work pool:

    <div class="terminal">

    ```bash
    python repo_info.py
    ```

    </div>

> Prefect builds a custom Docker image containing your workflow code that the worker can use to dynamically spawn Docker containers whenever this workflow needs to run.

Reviewer: ```suggestion
Prefect builds a custom Docker image containing your workflow code that the worker can use to dynamically spawn Docker containers whenever this workflow runs.
```

### Modify the deployment

To modify your deployment, edit your script and run it again. 

## Customize infrastructure settings

> Customize infrastructure settings for specific deployments by updating the `job_variables` section:

Reviewer: The important point here is that you're overriding the work pools' defaults, so I'd leave that part in.

> * **Local testing**: Set `image_pull_policy: Never` to avoid pulling from registries.

Reviewer: ```suggestion
* **Local testing**: Set `image_pull_policy: Never` to avoid pulling from a registry.
```

> * **Production**: Always push images to a remote registry for reliability.

Reviewer: ```suggestion
* **Production**:  Push your image to a remote registry for access.
```

</div>

**Remember**: Store deployment scripts at the repository root to avoid missing files in the Docker image.

## Next steps

> - Explore deployment configurations with [`prefect.yaml`](/guides/prefect-deploy/).

Reviewer 2: ```suggestion
- Explore deployment configuration with [`prefect.yaml`](/guides/prefect-deploy/).
```

> - Explore deployment configurations with [`prefect.yaml`](/guides/prefect-deploy/).

Reviewer: ```suggestion
- Explore an alternative way to specify work pool-based deployments with YAML using a [`prefect.yaml`](/guides/prefect-deploy/) file.
```

> - Deep dive into Prefect [concepts](/concepts/).

Reviewer: ```suggestion
- Explore Prefect [concepts](/concepts/).
```

- Use Prefect [guides](/guides/) for:

    - [Deploying flows on Kubernetes](/guides/deployment/kubernetes/)

>     - [Deploying flows in Docker](/guides/deployment/docker/)

Reviewer 2: ```suggestion
    - [Deploying flows in Docker containers](/guides/deployment/docker/)
```
